### PR TITLE
Fix flaky UI tests (#2403)

### PR DESCRIPTION
**Pull request by evabishchevich**
## Description
 This PR provides a fix for UI tests. Tested locally, all 10 attempts were green

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### CODE REVIEW

File: examples/build.gradle [hunk 1]
```diff
@@ -90,7 +90,9 @@ dependencies {
     implementation project(':libnavigation-util')
     implementation project(':liblogger')
     implementation project(':libnavigator')
-    androidTestImplementation project(':libtesting')
+    androidTestImplementation(project(':libtesting')) {
+        exclude group: 'org.robolectric'
```
Comment (evabishchevich) [inline examples/build.gradle]: Excluded because of 
https://github.com/robolectric/robolectric/issues/5235

Comment (korshaknn) [reply examples/build.gradle]: in what modules do we have duplicates?
can It be fixed in `libtesting` module to avoid such fixes in every consumer module?

Comment (evabishchevich) [reply examples/build.gradle]: It's not the only problem.  
The second one - we got 'java.lang.ClassCastException: android.os.Looper cannot be cast to org.robolectric.internal.bytecode.ShadowedObject' when run UI tests on simulators/devices.  Don't know why runner is trying to make shadows of real classes

Comment (evabishchevich) [reply examples/build.gradle]: Alternatively we can create different configurations inside `libtesting` module and provide different  dependencies

Comment (korshaknn) [reply examples/build.gradle]: I'm confused why we didn't have such problems before(`duplicating` and `ClassCastException`)? There were no such problems in `driver` repo
Any ideas?

Comment (evabishchevich) [reply examples/build.gradle]: We didn't include `robolectric` in `androidTestImplementation` dependencies

File: libtesting-ui/build.gradle [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+        }
+    }
+}
+
+dependencies {
+    api project(':libtesting-utils')
+    // downgrade because of https://github.com/mockk/mockk/issues/281
+    api 'io.mockk:mockk-android:1.9.2'
```
Comment (LukasPaczos) [inline libtesting-ui/build.gradle]: Why does the instrumentation testing lib needs mocks?

File: libtesting-utils/build.gradle [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        }
+    }
+}
+
+dependencies {
+    api dependenciesList.androidxTestJunit
+    api(dependenciesList.barista) {
+        exclude group: 'org.jetbrains.kotlin'
+        exclude group: 'androidx.annotation'
+    }
```
Comment (LukasPaczos) [inline libtesting-utils/build.gradle]: This should only live in the ui testing lib.

File: libtesting-utils/build.gradle [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    api dependenciesList.junit
+    api dependenciesList.coroutinesTestAndroid
+
+    implementation dependenciesList.coroutinesAndroid
```
Comment (LukasPaczos) [inline libtesting-utils/build.gradle]: While this should only live in the unit test lib, shouldn't it?

File: examples/build.gradle (hunk 1 above)
Comment (Guardiola31337) [reply examples/build.gradle]: Don't we? 👀 https://github.com/mapbox/mapbox-navigation-android/blob/f223bc9d6dd33d6aaf622c512ccb35f454b441a2/libtesting/build.gradle#L34

Comment (evabishchevich) [reply examples/build.gradle]: @Guardiola31337  The libtesting module was added to `testImplementation`, not `androidTestImplementation` in other modules

**Comment by evabishchevich** [comment]
> What problem are we trying to solve with this new module setup? It feels we're deviating a bit of the ultimate goal which is to solve #2403
> 
> Also it seems we're going backwards to the setup we had before #2420
> 
> Can't we solve #2403 without introducing these new modules? As @korshaknn I'm also confused 😕 [#2463 (comment)](https://github.com/mapbox/mapbox-navigation-android/pull/2463#discussion_r377702951)

It's hard to do via one module because we need different setup for  unit and UI tests. https://github.com/mockk/mockk/issues/281, for instance, prevent us from using the same mockk version in both configs (for UI and unit).
Probably, it's possible to do in one module but it would involve different gradle configurations. Finally I don't think that it would be easier to understand/support different configs in comparison to different modules

File: libtesting-ui/build.gradle (hunk 2 above)
Comment (evabishchevich) [reply libtesting-ui/build.gradle]: We still may want to mock something  (i.e. our own dependencies) inside instrumentation tests. Ofc we don't use it for now, so we may add it later

File: libtesting-utils/build.gradle (hunk 3 above)
Comment (evabishchevich) [reply libtesting-utils/build.gradle]: agreed

**Comment by korshaknn** [comment]
@Guardiola31337, @LukasPaczos  
at the beginning, we added `libtesting` module with `ui` tests to driver's directory and everything worked fine. After that we moved the module to `navigation` repo. Some of modules inside navigation used `robolectric` as a delpendency. 
Our initial goal was to have all test dependencies inside `libtesting`,  so we moved `robolectric` and other test dependencies from all modules to `libtesting`. After that every module could just add `libtesting` as a dependency to access any test dependency. But at that point we started getting crashes. 
The problem is we can't wrap `ui` tests and `roboletric` with the same module and provide them as a single dependency (duplicating errors, ClassCastException, issues with testRunner)
Looks like it's easier to split tests for 2 modules `unit` and 'ui' and use them independent. But I'm not sure that everything will work fine if we add such 2 modules to `core` for example.
@evabishchevich Am I correct? please add some details if it's possible

**Comment by LukasPaczos** [comment]
It should work just fine @korshaknn, one will be added with `testImplementation` the other with `andtoidTestImplementation` and we shouldn't run into any conflicts.

**Comment by evabishchevich** [comment]
@korshaknn Everything should be fine  when we add 2 testing modules  to a `some module that contains ui and  logic` because we will use them for different configurations (`testImplementation` and `androidTestImplementation`)

**Comment by evabishchevich** [comment]
One more thing I want to discuss is a count of testing modules. Right now there are 3 testing modules. The structure is the following:
1) `libtesting-utils `(name is discussable). Here we're going to have shared code and dependencies. For now shared code is only one class `MainCoroutineRule`.
2) `libtesting`. This is for unit tests. For now contains `mockk` and `robolectric` dependencies.
3) `libtesting-ui`. Here we have custom barista rules and `barista` dependency respectively.

Alternatively we can make just 2 modules (`libtesting` and `libtesting-ui`) . 
To achieve that we 
1) need to move `MainCoroutineRule` to `libtesting` (probably we wouldn't ever need this inside instrumentation tests)
2) duplicate     
```
api dependenciesList.androidxTestJunit
api dependenciesList.junit
```
3) remove `libtesting-utils `

3 modules look better as a platform for growing, but I'm not sure that we're going to make a lot of testing extensions.
@LukasPaczos @Guardiola31337 @korshaknn What variant do you prefer?

**Comment by korshaknn** [comment]
I prefer to have 2 modules right now, I prefer to have as simple structure as possible until we can handle it.

**Comment by Guardiola31337** [comment]
> 1) need to move `MainCoroutineRule` to `libtesting` (probably we wouldn't ever need this inside instrumentation tests)

Until we introduce Coroutines in the UI 😉

In any case, we should [KISS](https://en.wikipedia.org/wiki/KISS_principle) as currently we don't need it ([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)).

> What variant do you prefer?

I was wondering if we can go even further and have only one module `libtesting-ui`. In other words, is it worth it to have a completely separate module just for [`MainCoroutineRule`](https://github.com/mapbox/mapbox-navigation-android/blob/master/libtesting/src/main/java/com/mapbox/navigation/testing/MainCoroutineRule.kt)? 🤔 What do you think? My point is that we shouldn't create a module just for "sharing" dependencies because we can always add those explicitly (`testImplementation` or `androidTestImplementation`) as we normally do.

Thinking more. What about having `libtesting-ui` and `libtesting-utils` including only `MainCoroutineRule` (for now) so it can be shared across modules and eventually used in the UI? Is that an option? What do you think @evabishchevich @korshaknn @LukasPaczos?

**Comment by evabishchevich** [comment]
> My point is that we shouldn't create a module just for "sharing" dependencies because we can always add those explicitly (`testImplementation` or `androidTestImplementation`) as we normally do.

As for me having module just to gather all needed dependencies is quite okay. As long as we develop SDK modules count will only increase, so it would be harder to maintain just copy pasted dependencies between 10+ modules.

>  In other words, is it worth it to have a completely separate module just for MainCoroutineRule?

I'd say that we have two modules right now to remove `robolectric` dependency. 
In future nothing prevents us from including `testing lib for unit tests` inside `instrumentation` tests, we just may have to exclude some conflicting dependencies.

@Guardiola31337  Does it make  sense?

**Comment by Guardiola31337** [comment]
Capturing from internal discussion with @evabishchevich that we're going to try the following:

- `libtesting-ui` 👉 Leave it as is i.e. as a separate module
- ~`libtesting`~ 👉 Remove `libtesting` module and if possible try to share dependencies using Gradle (instead of just creating a module for sharing dependencies) and if not add those dependencies explicitly (as we do normally)
- `libtesting-utils` 👉 Create `libtesting-utils` module and place `MainCoroutineRule` (for now) so it's not duplicated and can be used across different modules (including potentially UI if needed in the future)

cc @korshaknn @LukasPaczos

File: libtesting-utils/build.gradle [hunk 5]
```diff
@@ -0,0 +1,27 @@
+apply plugin: 'com.android.library'
+apply plugin: 'kotlin-android'
+apply plugin: 'kotlin-android-extensions'
```
Comment (RingerJK) [inline libtesting-utils/build.gradle]: not necessary 
```suggestion

```

File: libdirections-hybrid/build.gradle [hunk 6]
```diff
@@ -40,7 +40,8 @@ dependencies {
     implementation dependenciesList.kotlinStdLib
     implementation dependenciesList.coroutinesAndroid
 
-    testImplementation project(":libtesting")
+    testImplementation project(':libtesting-utils')
+    apply from: "${rootDir}/gradle/testing-dependencies.gradle"
```
Comment (Guardiola31337) [inline libdirections-hybrid/build.gradle]: ❤️

Comment (Guardiola31337) [reply libdirections-hybrid/build.gradle]: Maybe calling it `unit-testing-dependencies`? What do you think @evabishchevich?

**Pull request merged**